sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/FailureSafeParser.scala

```diff
@@ -15,50 +15,57 @@
  * limitations under the License.
  */
 
-package org.apache.spark.sql.execution.datasources
+package org.apache.spark.sql.catalyst.util
 
 import org.apache.spark.SparkException
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
-import org.apache.spark.sql.catalyst.util._
-import org.apache.spark.sql.internal.SQLConf
-import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.types.{DataType, StructType}
 import org.apache.spark.unsafe.types.UTF8String
 
 class FailureSafeParser[IN](
```

Comment:
instead of changing this file, can we update `JsonToStructs.parser`, to wrap array or map with a struct before creating `FailureSafeParser`?

Reply:
Frankly speaking I don't fully understand the idea. Let's look at an example. We should parser JSON arrays (one array per row) like:
```
[1, 2, 3]
[4, 5]
```
and an user provided the schema `ArrayType(IntegerType, true)`. So, you propose to wrap the array by `StructType(Seq(StructField(ArrayType(IntegerType, ...))))`, right? And use the code inside of `JacksonParser` which we disabled by `allowArrayAsStructs` for now?

Reply:
It's not about the real business logic, but an idea to make the code simpler.
```
val parserSchema = nullableSchema match {
  case s: StructType => s
  case other => new StructType("value", other)
}
new FailureSafeParser[UTF8String](
      input => rawParser.parse(input, createParser, identity[UTF8String]),
      mode,
      parserSchema,
      parsedOptions.columnNameOfCorruptRecord,
      parsedOptions.multiLine)
```

Then we don't need to change `FailureSafeParser`